### ISSUE TITLE
Disabling running converter/validator on setattr in user-provided init

### ISSUE THREAD

**Issue by zhiltsov-max** [python-attrs/attrs#920]
> > Also, is it possible to use "plain" `setattr` (without side-effects) just in the user-provided constructor? The use case is that we'd like to have a more performant implementation of `__init__` (with less function calls, because they start to hit), but still support checks on assignments after initialization.

>The generated `__init__` already _does_ use `object.__setattr__`, however it also runs the validators by hand. This is the original behavior. :)

> Someone asked about disabling validators before but it seems a weird feature; I think a thread-safe version of `with attrs.validators.disabled():` would be nicer, such that you can instantiate with validated data anytime (#187).

> That said, I think using immutable classes might be a better approach.

> Other than that, pls don't hijack other issues, this might become hard to follow. Feel free to open an own issue any time!

> _Originally posted by @hynek in https://github.com/python-attrs/attrs/issues/917#issuecomment-1027639590_

Hi, let's continue it a separate thread then. Basically, we'd like to improve performance, when there is a ~1M of objects created:

```python
import timeit
from attrs import field, define
from attrs.setters import NO_OP
from attrs.validators import disabled

def myconv(v):
    return v

def myvalid(a, b, v):
    assert isinstance(v, int)

@define
class A:
    x = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x):
        with disabled():
            self.x = 10

@define
class B:
    x = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x):
        self.x = 10

@define(on_setattr=NO_OP)
class C:
    x = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x):
        assert isinstance(x, int)
        self.x = 10

@define
class D:
    x = field(default=5, converter=myconv, validator=myvalid)

print(timeit.timeit('A(20)', globals=globals(), number=1000000))
print(timeit.timeit('B(20)', globals=globals(), number=1000000))
print(timeit.timeit('C(20)', globals=globals(), number=1000000))
print(timeit.timeit('D(20)', globals=globals(), number=1000000))
```

With this sample, I get the following times:

```
2.216905194334686
0.7810841575264931
0.15891999378800392
0.48295993357896805
```

With more checks ang args passed, it works even slower. So it doesn't look like using `with disabled()` is a good choice. The situation `C` looks attractive, but `NO_OP` it also disables checks on regular assignments after class construction.

**Comment by hynek** [comment]
Wait if you have your own `__init__`, the solution is simple: don’t use `self.x = x` but `object.__setattr__(self, “x”, x)`. That’s what attrs uses internally and circumvents all mechanisms (including freezing).

Depending on the number of attributes, it can be worth to cache `object.__setattr__` within `__init__` into a local variable to save a global lookup + attribute lookup.

**Comment by zhiltsov-max** [comment]
Yeah, thanks for the idea! I've slightly extended the example to make it more realistic and tested it:

<details>

```python
import timeit
from attrs import field, define
from attrs.setters import NO_OP
from attrs.validators import disabled

def myconv(v):
    return v

def myvalid(a, b, v):
    assert isinstance(v, int)

@define
class A:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x, y, z, v, w):
        with disabled():
            self.x = x
            self.y = y
            self.z = z
            self.v = v
            self.w = w

@define
class B:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x, y, z, v, w):
        self.x = x
        self.y = y
        self.z = z
        self.v = v
        self.w = w

@define(on_setattr=NO_OP)
class C:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x, y, z, v, w):
        assert isinstance(x, int)
        self.x = x

        assert isinstance(y, int)
        self.y = y

        assert isinstance(z, int)
        self.z = z

        assert isinstance(v, int)
        self.v = v

        assert isinstance(w, int)
        self.w = w

@define(on_setattr=NO_OP)
class C2:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x, y, z, v, w):
        assert isinstance(x, int)
        self.x = myconv(x)

        assert isinstance(y, int)
        self.y = myconv(y)

        assert isinstance(z, int)
        self.z = myconv(z)

        assert isinstance(v, int)
        self.v = myconv(v)

        assert isinstance(w, int)
        self.w = myconv(w)

@define
class D:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

@define
class E:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x, y, z, v, w):
        object.__setattr__(self, "x", x)
        object.__setattr__(self, "y", y)
        object.__setattr__(self, "z", z)
        object.__setattr__(self, "v", v)
        object.__setattr__(self, "w", w)

@define
class E2:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x, y, z, v, w):
        _setattr = object.__setattr__
        _setattr(self, "x", x)
        _setattr(self, "y", y)
        _setattr(self, "z", z)
        _setattr(self, "v", v)
        _setattr(self, "w", w)

@define
class E22:
    x = field(default=5, converter=myconv, validator=myvalid)
    y = field(default=5, converter=myconv, validator=myvalid)
    z = field(default=5, converter=myconv, validator=myvalid)
    v = field(default=5, converter=myconv, validator=myvalid)
    w = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self, x, y, z, v, w):
        _setattr = object.__setattr__
        _setattr(self, "x", myconv(x))
        _setattr(self, "y", myconv(y))
        _setattr(self, "z", myconv(z))
        _setattr(self, "v", myconv(v))
        _setattr(self, "w", myconv(w))

print(timeit.timeit('A(20, 2, 3, 4, 5)', globals=globals(), number=1000000))
print(timeit.timeit('B(20, 2, 3, 4, 5)', globals=globals(), number=1000000))
print(timeit.timeit('C(20, 2, 3, 4, 5)', globals=globals(), number=1000000))
print(timeit.timeit('C2(20, 2, 3, 4, 5)', globals=globals(), number=1000000))
print(timeit.timeit('D(20, 2, 3, 4, 5)', globals=globals(), number=1000000))
print(timeit.timeit('E(20, 2, 3, 4, 5)', globals=globals(), number=1000000))
print(timeit.timeit('E2(20, 2, 3, 4, 5)', globals=globals(), number=1000000))
print(timeit.timeit('E22(20, 2, 3, 4, 5)', globals=globals(), number=1000000))

>>>
4.275309996999567 attrs + "disabled" cm
3.104956644005142 attrs + explicit init
0.3418383249954786 plain + checks
0.49537947998032905 plain + checks + conv
1.2905472760030534 attrs
0.6376119140186347 object.setattr raw
0.5214067950146273 local setattr raw
0.7132749549928121 local setattr + conv
```

</details>

The local setattr looks more or less good and it's only ~50% worse than the plain one. Quite interesting that case D worked 2 times faster than B - great work! I can't say I really satisfied by performance of such a simple example, but it is more to the interpreter performance itself. It looks like it's almost maximum we can get with this approach, so probably, we can consider the question resolved.

**Comment by hynek** [comment]
Cool! I'm always happy to see people take performance seriously and not just talk about it.
